File: src/main/scala-old/com/ajjpj/adiagram_/render/base/PartialImage.scala
package com.ajjpj.adiagram_.render.base

import com.ajjpj.adiagram_.geometry.{ARect, APoint}
import javafx.scene.image.Image
import javafx.scene.canvas.{Canvas, GraphicsContext}
import com.ajjpj.adiagram_.geometry.transform.Translation
import com.ajjpj.adiagram_.render.RenderHelper
import com.ajjpj.adiagram_.ui.fw.Digest
import com.ajjpj.adiagram_.ui.{Zoom, AScreenPos}


/**
 * This is the result of a rendering operation of a shape. To save memory (and improve cacheability), it does
 *  not contain the entire canvas but rather an arbitrary part of it plus a coordinate offset for rendering.
 *
 * @param renderOffset are the coordinates at which the upper left corner of the image should be rendered
 *
 * @author arno
 */
class PartialImage (val renderOffset: APoint, val img: Image)

class PartialImageWithShadow(val shape: PartialImage, val shadow: Option[PartialImage])

object PartialImage {
  type PartialImageRenderCallback = (GraphicsContext, Translation) => Unit

  /**
   * @param bounds of the actual image (toModel shadow, if any)
   * @param shadowStyle null if there is no shadow
   */
  def fromGc(pos: APoint, bounds: ARect, zoom: Zoom, shadowStyle: Option[ShadowStyle], callback: PartialImageRenderCallback): PartialImageWithShadow = {
    // one pixel on each side for sub-pixel bleeding
    val offset = bounds.topLeft + ((-1/zoom.factor, -1/zoom.factor))

    val canvas = new Canvas(bounds.width*zoom.factor+2, bounds.height*zoom.factor+2)
    val t = Translation(offset.inverse)
    callback(canvas.getGraphicsContext2D, t)

    val img = RenderHelper.snapshot(canvas)

    shadowStyle match {
      case Some(s) => new PartialImageWithShadow(new PartialImage(offset - pos, img), Some(new PartialImage(s.withShadowOffset(offset) - pos, s.shadow(img, zoom))))
      case None    => new PartialImageWithShadow(new PartialImage(offset - pos, img), None)
    }
  }
}

File: src/main/kotlin/com/ajjpj/adiagram/render/PartialImage.kt
package com.ajjpj.adiagram.render

import com.ajjpj.adiagram.Zoom
import com.ajjpj.adiagram.geometry.APoint
import com.ajjpj.adiagram.geometry.ARect
import com.ajjpj.adiagram.geometry.AffineTransformation
import com.ajjpj.adiagram.render.style.ShadowStyle
import javafx.scene.canvas.Canvas
import javafx.scene.canvas.GraphicsContext
import javafx.scene.image.Image

/**
 * This is the result of a rendering operation of a shape. To save memory (and improve cacheability), it does
 *  not contain the entire canvas but rather an arbitrary part of it plus a coordinate offset for rendering.
 *
 * @param renderOffset are the coordinates at which the upper left corner of the image should be rendered
 *
 * @author arno
 */
class PartialImage (val renderOffset: APoint, val img: Image) {
    companion object {

        /**
         * @param bounds of the actual image (toModel shadow, if any)
         * @param shadowStyle null if there is no shadow
         */
        fun fromGc(pos: APoint, bounds: ARect, zoom: Zoom, shadowStyle: ShadowStyle?, callback: (GraphicsContext, AffineTransformation) -> Unit): PartialImageWithShadow {
            // one pixel on each side for sub-pixel bleeding
            val offset = bounds.topLeft + APoint(-1/zoom.factor, -1/zoom.factor)

            val canvas = Canvas(bounds.width*zoom.factor+2, bounds.height*zoom.factor+2)
            val t = AffineTransformation.translation(offset.inverse)
            callback(canvas.graphicsContext2D, t)

            val img = RenderHelper.snapshot(canvas)

            if(shadowStyle != null) {
                return PartialImageWithShadow(PartialImage(offset - pos, img), PartialImage(shadowStyle.withShadowOffset(offset) - pos, shadowStyle.shadow(img, zoom)))
            }
            else {
                return PartialImageWithShadow(PartialImage(offset - pos, img), null)
            }
        }
    }
}

class PartialImageWithShadow(val shape: PartialImage, val shadow: PartialImage?)
